### PR TITLE
refactor TenantState transitions

### PR DESCRIPTION
**Pull request by problame**
Motivation
==========

This is preliminary work for/from #4220 (async `Layer::get_value_reconstruct_data`).
The motivation is to avoid locking `Tenant::timelines` in places that can't be `async`, because in #4333 we want to convert Tenant::timelines from `std::sync::Mutex` to `tokio::sync::Mutex`.

But, the changes here are useful in general because they clean up & document tenant state transitions.
That also paves the way for #4350, which is an alternative to #4333 that refactors the pageserver code so that we can keep the `Tenant::timelines` mutex sync.

Changes
=======

This patch consists of the following core insights and changes:

* spawn_load and spawn_attach own the tenant state until they're done
* once load()/attach() calls are done ...
   * if they failed, transition them to Broken directly (we know that there's no background activity because we didn't call activate yet)
   * if they succeed, call activate. We can make it infallible. How? Later.
   * set_broken() and set_stopping() are changed to wait for spawn_load() / spawn_attach() to finish.
   * This sounds scary because it might hinder detach or shutdown, but actually, concurrent attach+detach, or attach+shutdown, or load+shutdown, or attach+shutdown were just racy before this PR.
     So, with this change, they're not anymore.
     In the future, we can add a `CancellationToken` stored in Tenant to cancel `load` and `attach` faster, i.e., make `spawn_load` / `spawn_attach` transition them to Broken state sooner.

See the doc comments on TenantState for the state transitions that are now possible.
It might seem scary, but actually, this patch reduces the possible state transitions.

We introduce a new state `TenantState::Activating` to avoid grabbing the `Tenant::timelines` lock inside the `send_modify` closure.
These were the humble beginnings of this PR (see Motivation section), and I think it's still the right thing to have this `Activating` state, even if we decide against async `Tenant::timelines` mutex. The reason is that `send_modify` locks internally, and by moving locking of Tenant::timelines out of the closure, the internal locking of `send_modify` becomes a leaf of the lock graph, and so, we eliminate deadlock risk.

Patch Stack
===========

This PR sits on top of

- #4319

because otherwise we'd need to deal with falliblity of `Timeline::activate()` inside the `if activating {` branch (e.g., stopping already started background loops, etc).

### CODE REVIEW

File: pageserver/src/tenant.rs
Comment (shanyp) [inline pageserver/src/tenant.rs:1658]: NIT: shouldn't this comment be part of start_background_loops itself?

File: pageserver/src/tenant.rs
Comment (shanyp) [inline pageserver/src/tenant.rs:1662]: wouldn't this one will be always not_broken_timelines.size() ?

File: pageserver/src/tenant.rs
Comment (shanyp) [inline pageserver/src/tenant.rs:1737]: could this be called from a map() method in the previous line ?

File: pageserver/src/tenant.rs
Comment (koivunej) [reply pageserver/src/tenant.rs:1662]: In #4319 it just became infallible. Might be ok to such changes out of that PR series.

File: pageserver/src/tenant.rs
Comment (koivunej) [reply pageserver/src/tenant.rs:1737]: With `.map` you'd have the problem of "iterators are lazy and do nothing yadda yadda". `.for_each(|tl| tl.set_state(TimelineState::Stopping))` exists for this reason.

File: pageserver/src/tenant.rs
Comment (problame) [reply pageserver/src/tenant.rs:1662]: Fixed in https://github.com/neondatabase/neon/pull/4319/commits/4001f441c06b9aee4c422133e08fbfa622449e51

File: pageserver/src/tenant.rs
Comment (problame) [reply pageserver/src/tenant.rs:1737]: `not_broken_timelines` is an iterator and we need the precise number of elements it yielded (for `activated_timelines`).

So, leaning towards keeping the for loop.

File: libs/pageserver_api/src/models.rs
Comment (koivunej) [inline libs/pageserver_api/src/models.rs:65]: if `Loading` is `Attached` I can't see any other way for `Activating` than to be `Attached`

File: pageserver/src/tenant.rs
Comment (koivunej) [reply pageserver/src/tenant.rs:1670]: I spent some time trying to find this thread going back between first and fourth PRs.

Iterators are consumed in `for _ in iter` loops nor do they ever know the exact len, so 4001f44 had to go.

File: pageserver/src/tenant.rs
Comment (koivunej) [inline pageserver/src/tenant.rs:1681]: `assert_eq!` will in all cases give out a better assertion failure, so better use it here, add PartialEq if needed ... Even though it does feel icky if missing for the Broken state, let's just tolerate that.

```suggestion
                assert_eq!(
                    *current_state,
                    TenantState::Activating,
                    "set_stopping and set_broken wait for us to leave Activating state",
                );
```

File: pageserver/src/tenant.rs
Comment (koivunej) [reply pageserver/src/tenant.rs:1745]: Funny how this thread is now here, explains why I failed to find it.

enumerate would just number each of the timelines with their index. Usually it feels icky to do modify state within `Iterator::map` and return the index, then force evaluation with `max` or leave the enumerate even and do a plain `fold`, so you can eventually get the count by `let count = not_broken_timelines....max().map(|index| index + 1).unwrap_or(0);`.

Luckily it seems no need for count anymore here.

File: pageserver/src/tenant.rs
Comment (koivunej) [inline pageserver/src/tenant.rs:1714]: This is exactly why I dislike using watch channels.

From the `Receiver::borrow(&self)` docs:

> This method does not mark the returned value as seen, so future calls to [changed](https://docs.rs/tokio/latest/tokio/sync/watch/struct.Receiver.html#method.changed) may return immediately even if you have already seen the value with a call to borrow.

I think we should instead:

```suggestion
        while *rx.borrow_and_update() == TenantState::Activating {
            rx.changed()
                .await
                .expect("we're a method on Tenant, so, we're keeping self.state alive here");
        }
```

With `Receiver::borrow_and_update`:

> This method marks the current value as seen. Subsequent calls to [changed](https://docs.rs/tokio/latest/tokio/sync/watch/struct.Receiver.html#method.changed) will not return immediately until the [Sender](https://docs.rs/tokio/latest/tokio/sync/watch/struct.Sender.html) has modified the shared value again.

Related part about `Receiver::changed`:

> If the newest value in the channel has not yet been marked seen when this method is called, the method marks that value seen and returns immediately. If the newest value has already been marked seen, then the method sleeps until a new message is sent by the [Sender](https://docs.rs/tokio/latest/tokio/sync/watch/struct.Sender.html) connected to this Receiver, or until the [Sender](https://docs.rs/tokio/latest/tokio/sync/watch/struct.Sender.html) is dropped.

So maybe it's just a single extra iteration? Well, my confusion here is why I dislike watch with borrow + changed, but it with 1.28.1 now has better api available:

```suggestion
        {
            rx.wait_for(|state| state != TenantState::Activating)
                .await
                .expect("cannot drop self.state while on a &self method");
        }
```

Extra block for shortening the lifetime of the returned Ref type. I don't know how long it could get extended, but this way it surely will not be.

File: pageserver/src/tenant.rs
Comment (koivunej) [inline pageserver/src/tenant.rs:1708]: I'd prefer to drop this whole comment... It might even be outdated now?

No guaratees for observing every version with watch I think, ever, it could still hop multiple times ahead.

Might be worthwhile to refactor this `wait_for` (my suggestion below) out, and there add comments about why is this sane... I tried to come up with the state regex but it is not trivial to get it right in the time I have right now.

File: pageserver/src/tenant.rs
Comment (koivunej) [inline pageserver/src/tenant.rs:1748]: This reminded me of "don't ask, tell": we could just request all timelines to go stopping, and the broken ones could just ignore it?

File: libs/pageserver_api/src/models.rs
Comment (problame) [reply libs/pageserver_api/src/models.rs:65]: Hm, true, gotta think of this in light of https://neondb.slack.com/archives/C033A2WE6BZ/p1684866965026069?thread_ts=1684854161.739799&cid=C033A2WE6BZ as well

File: pageserver/src/tenant.rs
Comment (problame) [reply pageserver/src/tenant.rs:1714]: Nice, I didn't know there's `wait_for` , will switch

File: pageserver/src/tenant.rs
Comment (problame) [reply pageserver/src/tenant.rs:1748]: They do ignore it, see `Timeline::set_state` impl.
I think we're checking for `!TimelineState::Broken` so that we don't cause log noise?

IDK, it has been there before, I'm just moving the code.

File: pageserver/src/tenant.rs
Comment (problame) [reply pageserver/src/tenant.rs]: fe4ef121b6c4ffeaf9175001c463cfdc598164d1

File: pageserver/src/tenant.rs
Comment (problame) [reply pageserver/src/tenant.rs]: Dropped the comment & switched to wait_for in fe4ef121b6c4ffeaf9175001c463cfdc598164d1

File: pageserver/src/tenant.rs
Comment (koivunej) [reply pageserver/src/tenant.rs]: Oki understanding the original issue for `Receiver::wait_for` through the feature impl tokio PR 5611 (intentionally not linking), this usage of `borrow + changed` was correct after all. Would still prefer the common `self.wait_past_activating().await;` or something similar, where we could document the:

```
(attaching|loading)
  (activating
    (active
      (stopping (broken)?|broken)
    |broken)
  |broken)
```

I think the regex matches the drawnings, but it doesn't look good.

File: pageserver/src/tenant.rs
Comment (koivunej) [reply pageserver/src/tenant.rs:1746]: yeah I guess it's fine. resolving.

File: pageserver/src/tenant.rs
Comment (koivunej) [inline pageserver/src/tenant.rs:1716]: Thinking about is this really unreachable?

We wait_for state to not be activating, but I don't see this covering exhaustively.
| t   | first task loading | second task trying set_stopping      |
| --- | ------------------ | ------------------------------------ |
| t=0 | state=Loading      |                                      |
| t=1 |                    | check state!=Activating              |
| t=2 | state==Activating  |                                      |
| t=3 |                    | unreachable!("we checked above...")  |

Or what am I missing? Could be something obvious.

Comment (problame) [reply pageserver/src/tenant.rs:1716]: Right, need to move the stuff inside the `send_modify` closure into the `wait_for` closure.
Luckily, wait_for passes a mut ref

Comment (problame) [reply pageserver/src/tenant.rs:1716]: Got an idea

Comment (koivunej) [reply pageserver/src/tenant.rs:1716]: My idea was to just loop back if we find ourselves in the current `unreachable` within `send_modify`.

File: libs/pageserver_api/src/models.rs
Comment (koivunej) [reply libs/pageserver_api/src/models.rs:65]: So right now "not found from s3" tenant on attach will yield broken and is assumed to *stay* broken, as in we will not get rid of the tenant in memory? Confirmed we don't.

For me the answer seems to be quite simple; just forget everything about the tenant if we never had any possibility of attaching it? Problem will probably be then crash-safety, because we will have the files on disk. It seems even the attach marker was left, so until a manual detach comes we can't make progress even across restarts. This is most likely current design.

I am unsure if we *need* to focus on solving the "tried to attach tenant from different region" issue.

**Comment by hlinnaka** [comment]
> To enable that conversion, we must only lock Tenant::timelines in places that are already async or can be turned async.

Well, there's also [RwLock::blocking_read](https://docs.rs/tokio/latest/tokio/sync/struct.RwLock.html#method.blocking_read).

 > The closure passed to Tenant::state.send_modify() is the only really tricky place that doesn't allow async, so, this PR refactors the code to move the locking outside of the closure.

Hmm, yes those send_modify() closures are a little scary before this PR. You're really not supposed to do much inside the send_modify() closure, because it holds the lock on the watch's value. But we acquire the `timelines` lock and do a fair amount of processing on each timeline. It doesn't feel right. This PR helps with that, but I'm not sure if the cure is worse than the disease, because now all the readers of the watch have an extra state to worry about.

**Comment by problame** [comment]
Pushed my solution into this branch, I think it's best to not view at it in isolation but instead just re-review this entire PR https://github.com/neondatabase/neon/pull/4321/commits/de780d2e0fd0718522b5011dffc4a6f96ccc0628

File: pageserver/src/tenant.rs
Comment (koivunej) [reply pageserver/src/tenant.rs]: Oki I see now my `set_stopping` should had been `set_broken`.

File: libs/pageserver_api/src/models.rs
Comment (koivunej) [inline libs/pageserver_api/src/models.rs:43]: This can be viewed with https://mermaid.live at least. Unsure if this is the way to go.

Comment (problame) [reply libs/pageserver_api/src/models.rs:43]: > Well, apparently support exists: https://github.com/mersinvald/aquamarine

Yeah, I was afraid of the review comments for pulling that in ;)

File: pageserver/src/tenant.rs
Comment (koivunej) [inline pageserver/src/tenant.rs:1762]: ok so the commit contrains how can we enter broken, so for example set_broken during fail-point paused load/activate would not be allowed but only right after.

Yeah, it does not sound so useful anymore.

File: pageserver/src/tenant/mgr.rs
Comment (koivunej) [inline pageserver/src/tenant/mgr.rs:578]: Should this be pulled out of holding TENANTS lock?

File: pageserver/src/tenant.rs
Comment (koivunej) [reply pageserver/src/tenant.rs]: Oki, de780d2 looks quite ok to me. I do hope we don't get other ideas of setting tenants to broken than the ones we currently have in tests. Luckily those are `active -> broken`, so enabling that for non-testing builds does not sound like a huge chore.

Resolving.

File: libs/pageserver_api/src/models.rs
Comment (koivunej) [reply libs/pageserver_api/src/models.rs:43]: Also regex version would work and not require an external thing to view.

File: pageserver/src/tenant.rs
Comment (koivunej) [reply pageserver/src/tenant.rs:1762]: resolving.

File: pageserver/src/tenant.rs
Comment (koivunej) [inline pageserver/src/tenant.rs:1691]: logging reminder; I think it could be made just inline here:

```suggestion
            TenantState::Activating | TenantState::Loading | TenantState::Attaching => {
                info!("waiting for {} to turn Active|Broken|Stopping", <&'static str>::from(state));
                false
            }
```

Should make sure we never log the full states because they can contain the stacktrace, but I haven't reminded myself of in which they contain or dont.

Comment (koivunej) [reply pageserver/src/tenant.rs:1691]: I dislike strum macros once more. At least the `<&'static str>::from(state)` will not contain a stacktrace :)

File: libs/pageserver_api/src/models.rs
Comment (problame) [reply libs/pageserver_api/src/models.rs:43]: What is this "regex version" you are talking about?

File: pageserver/src/tenant/mgr.rs
Comment (problame) [reply pageserver/src/tenant/mgr.rs:578]: It's needed for the `wait_for`.

File: libs/pageserver_api/src/models.rs
Comment (koivunej) [reply libs/pageserver_api/src/models.rs:43]: Before your changes: https://github.com/neondatabase/neon/pull/4321#discussion_r1205238263

File: pageserver/src/tenant/mgr.rs
Comment (koivunej) [reply pageserver/src/tenant/mgr.rs:595]: > needed for the `wait_for`

What do you mean? 

`set_stopping(&self)`, so the tenant could be smuggled out of this block, and then the await done outside of the write lock on TENANTS.

File: libs/pageserver_api/src/models.rs
Comment (problame) [reply libs/pageserver_api/src/models.rs]: pushed https://github.com/neondatabase/neon/pull/4321/commits/1367e2b0eebc08a9add60f087a5024014ab4d7a5

File: pageserver/src/tenant.rs
Comment (problame) [reply pageserver/src/tenant.rs]: Applied your suggestion to both wait_for calls, thanks!

File: libs/pageserver_api/src/models.rs
Comment (problame) [reply libs/pageserver_api/src/models.rs:106]: I'll put this in the right issue

Comment (problame) [reply libs/pageserver_api/src/models.rs:106]: https://github.com/neondatabase/neon/issues/4344#issuecomment-1563947761

**Comment by koivunej** [comment]
> > To enable that conversion, we must only lock Tenant::timelines in places that are already async or can be turned async.
> 
> Well, there's also [RwLock::blocking_read](https://docs.rs/tokio/latest/tokio/sync/struct.RwLock.html#method.blocking_read).

`blocking_read` uses `block_on` which cannot be used from within async context ([heading panics](https://docs.rs/tokio/latest/tokio/sync/struct.RwLock.html#panics-1)). "Async context" meant that you are anywhere called by an async task from tokio async workers.

> > The closure passed to Tenant::state.send_modify() is the only really tricky place that doesn't allow async, so, this PR refactors the code to move the locking outside of the closure.
> 
> Hmm, yes those send_modify() closures are a little scary before this PR. You're really not supposed to do much inside the send_modify() closure, because it holds the lock on the watch's value. But we acquire the `timelines` lock and do a fair amount of processing on each timeline. It doesn't feel right. This PR helps with that, but I'm not sure if the cure is worse than the disease, because now all the readers of the watch have an extra state to worry about.

I don't remember which PR it was, perhaps not yet a PR, but this problem of observing `TenantState` could/should be solved by instead of relying externally on `TenantState`, rely on  the fact that tenant is now active so we start background jobs. We can only stop from being active by being stopping or broken.

**Review by koivunej — APPROVED** [review]
Pre-emptively approving as this is looking good with concerns noted in open threads:
- I don't see why me must lock global `TENANTS.write()` while we transition to stopping -- I think it could be handled by doing the shutdown only in the task which won the set_stopping transition, as there can be only one. At minimum we should document that `TENANTS.write()` is only locked for request coalescing, if that is the rationale.
- mermaid => regex or rustdoc compatible

Additional thread kept unresolved for regex example.

File: pageserver/src/tenant/mgr.rs
Comment (problame) [reply pageserver/src/tenant/mgr.rs:595]: > What do you mean as in how is it needed? I did read the comment above, but it seems this is at most a proxy for request coalescing. The proxy is just a global RWLock :)
> 
> set_stopping(&self), so the tenant could be smuggled out of this block, and then the await done outside of the write lock on TENANTS, pick the one who manages to make the state transition as the one and lucky?
> 

Hm, I can't really follow, maybe best to leave these two paragraphs alone and focus on the third.

> Well, perhaps this could be in a FIXME just as well. I am just worrying that the wait_for might loop for a long time in some unforseen situation.

Let's look at the remaining callers of `set_stopping` and `set_broken`. 

* `shutdown_all_tenants -> set_stopping` => implicit timeout through the SIGKILL. It might make sense to do FuturesUnordered there? WDYT?
* `remove_tenant_from_memory` => this is detach, which currently has cancel-safety prolems[^1]. I think the `wait_for` doesn't change anything here.
* `remove_tenant_from_memory` => `set_broken` is only used if the cleanup future fails, and, it's the last call on the way out. So, I don't expect breakage.


[^1]: can't find the issue, but, it's the one where if client hangs up during detach, we fail because tenant is already `Stopping`

Comment (LizardWizzard) [reply pageserver/src/tenant/mgr.rs:595]: Holding write lock on TENANTS map will prevent new page_service connections from being established (globally, not only for this particular tenant), which I think is undesirable

**Comment by problame** [comment]
> mermaid => regex or rustdoc compatible

I like that I can annotate the edges in mermaid notation though. The regex notation doesn't have that / would become bloated if it had.

> I don't see why me must lock global TENANTS.write() while we transition to stopping -- I think it could be handled by doing the shutdown only in the task which won the set_stopping transition, as there can be only one

Good find, I had completely missed that problem with the `wait_for` in `set_stopping`.

Yeah, that proposal sounds reasonable, I'll push sth.

File: pageserver/src/tenant/mgr.rs
Comment (problame) [reply pageserver/src/tenant/mgr.rs]: fixed in https://github.com/neondatabase/neon/pull/4321/commits/e7c4ef9f4f8ac3e5e5a383839ba0b365e4e6039b and preceding commit

File: pageserver/src/tenant.rs
Comment (koivunej) [reply pageserver/src/tenant.rs]: Resolving, mermaid is not a reason to hold this PR back.

File: libs/pageserver_api/src/models.rs
Comment (koivunej) [reply libs/pageserver_api/src/models.rs:43]: I don't want to to block this PR on mermaid.

**Comment by koivunej** [comment]
Took me a while to find this PR again with the changed title. I'll peek at the latest changes next.

**Comment by problame** [comment]
Again, I would recommend https://github.com/orgs/neondatabase/projects/38/views/6 for a view on all the in-flight PRs.

File: pageserver/src/tenant.rs
Comment (problame) [reply pageserver/src/tenant.rs:1773]: It definitely can, but, `send_if_modified` requires returning a `bool` inside the closure anyways, so, might as well consistency-check it.

File: pageserver/src/tenant/mgr.rs
Comment (koivunej) [inline pageserver/src/tenant/mgr.rs]: Might not even require much, but marking TODO.

Comment (problame) [reply pageserver/src/tenant/mgr.rs]: It's fixed in the subsequent commit, already in this branch. Seems like you're reviewing commit by commit

File: pageserver/src/tenant.rs
Comment (koivunej) [reply pageserver/src/tenant.rs:1773]: Ah yes, missed the `send_if_modified` requirement. Perhaps resolving this then.

File: pageserver/src/tenant/mgr.rs
Comment (koivunej) [inline pageserver/src/tenant/mgr.rs:282]: The hook will have already reported this.

The only possible thing missing is a span around join_set spawned task, which could let the tenant_id be known, but for this site, it doesn't matter.

`set_stopping` does not have `#[instrument]`, I don't have the callers of shutdown_pageserver right now but I am quite certain they are not or can not set a useful span for this panick.

So, I'll patch this to be for example a counter an a single warning, then add the `.instrument(...)` for each of the `join_set.spawn`'d futures.

File: pageserver/src/tenant/mgr.rs
Comment (koivunej) [reply pageserver/src/tenant/mgr.rs]: Ah, TODO got implemented already.

File: pageserver/src/tenant/mgr.rs
Comment (problame) [reply pageserver/src/tenant/mgr.rs:282]: Ok, I'll leave that to you.

Comment (koivunej) [reply pageserver/src/tenant/mgr.rs]: 2cbb593d (fix this conv up) and 745e3d45 (adds `#[instrument]`s)

File: pageserver/src/tenant/mgr.rs
Comment (koivunej) [inline pageserver/src/tenant/mgr.rs:331]: I'll do this next, I don't think it's plausible to wait each one-by-one.

File: pageserver/src/tenant/mgr.rs
Comment (koivunej) [reply pageserver/src/tenant/mgr.rs]: 745e3d4 forgot an `use`...

Comment (koivunej) [reply pageserver/src/tenant/mgr.rs]: [6bee7df](https://github.com/neondatabase/neon/pull/4321/commits/6bee7df00b0cc69fcc3b6fa4b0b53b20dfb84e67) and [71fa6d9](https://github.com/neondatabase/neon/pull/4321/commits/71fa6d924ee148c9d199729ad03de2c5ceb3d80e) after a bit of touchup.

File: pageserver/src/tenant/mgr.rs
Comment (koivunej) [reply pageserver/src/tenant/mgr.rs]: [f46121e](https://github.com/neondatabase/neon/pull/4321/commits/f46121e869ed6bb8fa70dea407b29b3223764523)

File: pageserver/src/tenant/mgr.rs
Comment (koivunej) [inline pageserver/src/tenant/mgr.rs:253]:
```suggestion
    // Set tenant (and its timlines) to Stoppping state.
    //
    // Since we can only transition into Stopping state after activation is complete,
    // run it in a JoinSet so all tenants have a chance to stop before we get SIGKILLed.
```

File: pageserver/src/tenant/mgr.rs
Comment (koivunej) [inline pageserver/src/tenant/mgr.rs:258]:
```suggestion
    // 3. Signal cancellation for other tenant background loops.
```

File: pageserver/src/tenant/mgr.rs
Comment (koivunej) [reply pageserver/src/tenant/mgr.rs]: I still find the whole `TENANTS.write()` locking questionable, but its difficult to say if it's a load bearing exclusive lock. Resolving anyways.

**Review by problame — COMMENTED** [review]
changes on top lgtm, @koivunej please rerun tests to pass and merge, or get another review from third party monday

File: pageserver/src/tenant/mgr.rs
Comment (koivunej) [reply pageserver/src/tenant/mgr.rs:343]: Reporting panics via tracing is the 101, so I don't feel like adding this, but I do feel that the block shouldn't be empty.

I'd unify the two `join_error.is_panic()` arms for both join_sets but I'd rather have the tests run now as they started already.

**Comment by koivunej** [comment]
Yep, thanks. I don't think those latest changes are interesting enough. I'll approve next.

**Review by koivunej — APPROVED** [review]

**Comment by koivunej** [comment]
It seems there's an actual test failure:


test_timeline_resurrection_on_attach[True-local_fs]: [debug](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-4321/5109666122/index.html#suites/f588e0a787c49e67b29490359c589fae/36cbfeab70e5312c) with:

```
AssertionError: assert not ['2023-05-29T07:51:54.527416Z ERROR request{method=GET path=/v1/tenant/bdc883cb95af183d3c72095b938eeb95 request_id=2098afa1-e542-4213-acae-94ff27cfd86f}:tenant_status_handler{tenant=bdc883cb95af183d3c72095b938eeb95}:panic{thread=mgmt request worker location=libs/pageserver_api/src/models.rs:106:33}: not yet implemented\n']
```

Well, this doesn't seem like something changed in this PR, will investigate.

File: libs/pageserver_api/src/models.rs
Comment (koivunej) [reply libs/pageserver_api/src/models.rs:106]: Unresolved because this is already an issue on this PR: https://github.com/neondatabase/neon/pull/4321#issuecomment-1566729898

Comment (koivunej) [reply libs/pageserver_api/src/models.rs:106]: IMHO we could just set this to `Attached` right away, because we do require the manual detach before this. If we ever slim down the amount of reasons for `Attached`, it would be fine as well.

Comment (koivunej) [reply libs/pageserver_api/src/models.rs]: Solved as "attachment_status is same as the state where we arrived to Activating" in [d6d0c74](https://github.com/neondatabase/neon/pull/4321/commits/d6d0c74c0699a3ff0c7c977a4c46c29754c589f0)

Comment (LizardWizzard) [reply libs/pageserver_api/src/models.rs]: Green light from me

Comment (koivunej) [reply libs/pageserver_api/src/models.rs]: Thanks. CI is taking forever again. Resolving this.

**Pull request merged**
